### ISSUE TITLE
4. JEPA-Style Joint Embedding Predictive Architecture

### ISSUE THREAD

**Issue by ricardoamartinez** [ricardoamartinez/koopman-physics-neural-cellular-automata#5]
# JEPA-Style Joint Embedding Predictive Architecture

## Overview
Implement **Joint Embedding Predictive Architecture (JEPA)** training mechanism to learn compact abstract representations through contrastive embedding losses rather than direct pixel-wise reconstruction. This trains the network to predict parts of its input or future states in latent space, encouraging predictive relationship learning.

### Core JEPA Components
Define encoder and predictor modules for temporal prediction in latent space:

```python
class Encoder(nn.Module):
    def __init__(self, input_dim, latent_dim):
        super(Encoder, self).__init__()
        self.net = nn.Sequential(
            nn.Linear(input_dim, 256),
            nn.ReLU(),
            nn.Linear(256, latent_dim)
        )
    
    def forward(self, x):
        return self.net(x)

class PredictorHead(nn.Module):
    def __init__(self, latent_dim):
        super(PredictorHead, self).__init__()
        # Simple MLP to predict future embedding from current embedding
        self.net = nn.Sequential(
            nn.Linear(latent_dim, latent_dim)
            # Could add more layers or nonlinearity if needed
        )
    
    def forward(self, z):
        return self.net(z)
```

### InfoNCE Contrastive Loss Implementation
Implement contrastive loss using cosine similarity and temperature scaling:

```python
def contrastive_loss(pred, target, temperature=0.1):
    # pred, target: tensors of shape [batch, latent_dim]
    # Compute normalized cosine similarity
    pred_norm = F.normalize(pred, dim=1)
    target_norm = F.normalize(target, dim=1)
    
    # Positive cosine similarity (diagonal of pred vs target)
    pos_sim = (pred_norm * target_norm).sum(dim=1)  # batch of cosines
    
    # We need negatives: use other samples in batch as negatives (typical InfoNCE)
    # Compute full cosine similarity matrix between pred and target
    sim_matrix = torch.matmul(pred_norm, target_norm.T)  # shape [batch, batch]
    
    # Apply temperature and softmax for InfoNCE
    sim_matrix /= temperature
    
    # For each sample i, compute InfoNCE loss:
    # -log(exp(sim[ii]/T) / sum_j exp(sim[ij]/T))
    batch_size = pred.size(0)
    labels = torch.arange(batch_size, device=pred.device)
    loss = F.cross_entropy(sim_matrix, labels)
    return loss
```

### 1. Context-Target Pair Generation
Generate training pairs for temporal prediction:

```python
def generate_jepa_pairs(observations, k_steps=1):
    """
    Generate context-target pairs for JEPA training
    observations: list of sequential observations from simulation
    k_steps: prediction horizon (1 for next-step prediction)
    """
    context_obs = []
    target_obs = []
    
    for i in range(len(observations) - k_steps):
        context_obs.append(observations[i])       # current observation
        target_obs.append(observations[i + k_steps])  # future observation
    
    return context_obs, target_obs
```

#### Option A: Direct Latent Encoding
If NCA latent state already serves as good encoding:
```python
# Use NCA latent directly as context encoding
context_latent = nca_latent_t  # from NCA at time t
target_latent = nca_latent_t_k  # from NCA at time t+k
```

#### Option B: Separate Observation Encoder
For high-dimensional inputs (images):
```python
# Encode raw observations to latent space
context_embedding = encoder(obs_t.flatten())
target_embedding = encoder(obs_t_k.flatten()) 
```

### 3. Training Loop Integration
Integrate JEPA loss into the main training loop:

```python
# Within main simulation loop, if training_enabled:
if training_enabled and len(observation_buffer) > k_steps:
    # Sample context-target pairs
    context_obs, target_obs = generate_jepa_pairs(observation_buffer[-10:], k_steps=1)
    
    for ctx, tgt in zip(context_obs, target_obs):
        # Option 1: Use NCA rollout for context encoding
        ctx_latent = encoder(ctx.view(1, -1))
        ctx_latent = ctx_latent.view(1, channels, H, W)
        # Run NCA for k steps to get predicted future
        for _ in range(k_steps):
            ctx_latent = nca(ctx_latent)
        predicted_embedding = predictor_head(ctx_latent.view(1, -1))
        
        # Option 2: Direct encoding of target
        target_embedding = encoder(tgt.view(1, -1)).detach()
        
        # Compute JEPA contrastive loss
        jepa_loss = contrastive_loss(predicted_embedding, target_embedding)
        
        # Backpropagate
        optimizer.zero_grad()
        jepa_loss.backward()
        optimizer.step()
```

#### Masked Spatial Prediction
For static scenarios, predict masked regions:
```python
def create_spatial_mask(image, mask_ratio=0.5):
    """Create random spatial mask for image patches"""
    H, W = image.shape[-2:]
    mask = torch.rand(H, W) < mask_ratio
    masked_image = image.clone()
    masked_image[..., mask] = 0  # mask out random regions
    return masked_image, mask

# In training loop:
masked_obs, mask = create_spatial_mask(observation)
context_embedding = encoder(masked_obs.flatten())
target_embedding = encoder(observation.flatten())  # full observation
predicted_embedding = predictor_head(context_embedding)
loss = contrastive_loss(predicted_embedding, target_embedding)
```

#### Multi-Step Prediction
Predict multiple time steps ahead:
```python
def multi_step_jepa_loss(context_latent, target_sequence, predictor, k_max=5):
    """Compute JEPA loss over multiple prediction horizons"""
    total_loss = 0
    current_latent = context_latent
    
    for k in range(1, k_max + 1):
        # Predict k steps ahead
        predicted = predictor(current_latent)
        target = target_sequence[k-1]  # target at step k
        
        # Add weighted loss (closer predictions weighted higher)
        weight = 1.0 / k  # exponential decay of importance
        total_loss += weight * contrastive_loss(predicted, target)
        
        # Update latent for next prediction (teacher forcing)
        current_latent = predicted.detach()
    
    return total_loss
```

### VICReg-Style Loss
Alternative to InfoNCE using variance-invariance-covariance:

```python
def vicreg_loss(pred, target, sim_coeff=25.0, std_coeff=25.0, cov_coeff=1.0):
    """VICReg loss for JEPA training"""
    batch_size, dim = pred.shape
    
    # Invariance loss (similarity)
    sim_loss = F.mse_loss(pred, target)
    
    # Variance loss (prevent collapse)
    pred_std = torch.sqrt(pred.var(dim=0) + 1e-04)
    target_std = torch.sqrt(target.var(dim=0) + 1e-04)
    std_loss = torch.mean(F.relu(1 - pred_std)) + torch.mean(F.relu(1 - target_std))
    
    # Covariance loss (decorrelate features)
    pred_cov = torch.cov(pred.T)
    target_cov = torch.cov(target.T)
    cov_loss = torch.sum(torch.triu(pred_cov, diagonal=1)**2) + torch.sum(torch.triu(target_cov, diagonal=1)**2)
    
    total_loss = sim_coeff * sim_loss + std_coeff * std_loss + cov_coeff * cov_loss
    return total_loss
```

### Barlow Twins Approach
Cross-correlation matrix approach:

```python
def barlow_twins_loss(pred, target, lambda_off_diag=5e-3):
    """Barlow Twins loss for JEPA"""
    batch_size = pred.size(0)
    
    # Normalize features
    pred = F.normalize(pred, dim=0)
    target = F.normalize(target, dim=0)
    
    # Cross-correlation matrix
    cross_corr = torch.matmul(pred.T, target) / batch_size
    
    # Diagonal should be 1 (invariance), off-diagonal should be 0 (independence)
    on_diag = torch.diagonal(cross_corr).add_(-1).pow_(2).sum()
    off_diag = cross_corr.flatten()[1:].view(cross_corr.size(0)-1, cross_corr.size(0)+1)[:, :-1].pow_(2).sum()
    
    loss = on_diag + lambda_off_diag * off_diag
    return loss
```

## Acceptance Criteria
- [ ] Encoder maps observations to latent embeddings consistently
- [ ] PredictorHead produces meaningful future predictions
- [ ] InfoNCE loss correctly contrasts positive/negative pairs
- [ ] Integration with NCA preserves end-to-end differentiability
- [ ] Batch processing handles multiple context-target pairs efficiently
- [ ] Alternative loss variants (VICReg, Barlow Twins) implemented and tested
- [ ] Training convergence: loss decreases over time
- [ ] Learned representations enable better downstream task performance

### Memory Management
```python
# Detach target embeddings to prevent gradient accumulation
target_embedding = encoder(target_obs).detach()

# Use gradient checkpointing for long NCA rollouts
from torch.utils.checkpoint import checkpoint
predicted_latent = checkpoint(nca_rollout, context_latent, k_steps)
```

### Efficient Negative Sampling
```python
def efficient_negative_sampling(embeddings, num_negatives=10):
    """Sample negatives efficiently for large batch sizes"""
    batch_size = embeddings.size(0)
    if batch_size > num_negatives:
        # Randomly sample negatives instead of using all
        indices = torch.randperm(batch_size)[:num_negatives]
        return embeddings[indices]
    return embeddings
```

## Integration Notes
The JEPA training will be integrated in the main loop:

```python
# Instantiate JEPA components
encoder = Encoder(input_dim=observation_dim, latent_dim=128).to(device)
predictor_head = PredictorHead(latent_dim=128).to(device)

# Add to optimizer
jepa_params = list(encoder.parameters()) + list(predictor_head.parameters())
optimizer = torch.optim.Adam(list(nca.parameters()) + jepa_params, lr=1e-3)

# In main loop:
jepa_loss = compute_jepa_loss(current_obs, future_obs)
total_loss = jepa_loss + dsl_loss  # combine with other losses
```

This JEPA implementation enables the agent to learn **predictive latent representations** that capture temporal dynamics without requiring pixel-level reconstruction, leading to more robust and generalizable internal models.
